### PR TITLE
Docker runtime improvements

### PR DESCRIPTION
**Pull request by Wirone**
- Support for XDebug
- Reusable ESC/PHPStan cache for Docker runtime
- Bump memory limit for PHPStan inside Docker container

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -2,8 +2,13 @@ FROM php:8-cli-alpine
 
 WORKDIR /etc/rector
 
-# required for composer patches
-RUN apk add --no-cache patch
+RUN apk add --no-cache patch \
+    && apk add --no-cache --virtual .build-deps $PHPIZE_DEPS icu-dev \
+    && pecl install xdebug \
+    && docker-php-ext-enable xdebug \
```
Comment (TomasVotruba) [inline Dockerfile:8]: What about performance hit for normal Rector run? We had couple reports of slow Rector, then it took couple of weeks to find the default xdebug is to blame :) this cannot be here.

File: composer.json [hunk 2]
```diff
@@ -150,6 +150,7 @@
         "release": "vendor/bin/monorepo-builder release patch --ansi"
     },
     "extra": {
+        "enable-patching": true,
         "patches": {
```
Comment (TomasVotruba) [inline composer.json]: Could you send this in standalone PR? It's no brainer to merge 👍

File: ecs.php [hunk 3]
```diff
@@ -40,6 +40,9 @@
     // make ECS 16x faster
     $parameters->set(Option::PARALLEL, true);
 
+    // Reusable ECS cache for Docker runtime
+    $parameters->set(Option::CACHE_DIRECTORY, 'tmp/ecs');
+
     $parameters->set(Option::PATHS, [
```
Comment (TomasVotruba) [inline ecs.php]: The configs cannot be changed to make Docker happy, as the default way to develop Rector is natively. Same for PHPStan.

File: Dockerfile (hunk 1 above)
Comment (Wirone) [reply Dockerfile:8]: By default XDebug is transparent because of `XDEBUG_MODE` fallback value `off`. It must be enabled explicitly by setting `debug` in `.env` file. Extension enabled was a problem prior to v3, now it can be always enabled and there's no overhead.

File: ecs.php (hunk 3 above)
Comment (Wirone) [reply ecs.php]: But why? Default values make use of local system temp folder and in this scenario it's not possible to use cache inside Docker container (in fact, temp dir is used inside container but it disappears when container stops, so cache lives only through one `run` lifecycle and can't be reused). I don't see how it collides with non-Docker development 🤔 You'll just have cache in project's dir, not somewhere else in the filesystem.

File: Dockerfile (hunk 1 above)
Comment (TomasVotruba) [reply Dockerfile:8]: Coudl you link this knowledge in these lines? We're not using Docker ready, so we are aware of this explicitly

File: ecs.php (hunk 3 above)
Comment (TomasVotruba) [reply ecs.php]: It's our convention to narrow the debug use cases :). 
We used to follow different path before, adding more Docker-features to core for occasional contributors, but nobody used them and knew how to work with them. To make our lives easier and development more reliable, we've decide to follow one standard.

Comment (Wirone) [reply ecs.php]: Honestly, I don't understand this reasoning. For non-Docker usage it does not make difference where cache is stored, but for Docker it is really important because cache will work or not 🙂 Without specifying cache paths it becomes really hard to use because every run is like first run and takes much time. I believe nobody will even notice that cache moved to different directory, this change is 100% transparent for developers because nothing needs to be done in order to work. It just enables reusable cache inside Docker container.

File: composer.json (hunk 2 above)
Comment (Wirone) [reply composer.json]: Extracted to #1929

File: Dockerfile (hunk 1 above)
Comment (Wirone) [reply Dockerfile:8]: It's not Docker-related knowledge, it's how XDebug v3 works 😉 But I've added some info to `CONTRIBUTING.md`.

File: ecs.php (hunk 3 above)
Comment (TomasVotruba) [reply ecs.php]: It does matter actually, as clearing cache is related to this path etc.

Feel free to move it to Docker setup scripts :)

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -20,6 +22,13 @@ docker-compose run php composer install
 
 Now you can start using all scripts and work with the code.
 
+#### Using XDebug in Docker runtime
+
+XDebug is installed and enabled in Docker container, but by default it's in [`xdebug.mode=off`](https://xdebug.org/docs/all_settings#mode), adding 0 overhead. If you need to debug code, set `XDEBUG_MODE=debug` in `.env` file. You can tweak Docker config with these variables:
+
+- `XDEBUG_IDEKEY`: this must be the same as in IDE's remote debug configuration
+- `XDEBUG_CLIENT_HOST`: it's your host from container's perspective. By default, it's set to `host.docker.internal` which works for Docker Desktop on Mac. For Unix you can try with `172.16.0.1` or set `extra_hosts` in `docker-compose.override.yml` ([see here](https://github.com/docker/for-linux/issues/264#issuecomment-965465879)).
+
```
Comment (TomasVotruba) [inline CONTRIBUTING.md:49]: 👍

File: ecs.php (hunk 3 above)
Comment (Wirone) [reply ecs.php]: So clearing cache should be improved to be compatible 😉 Please point me to scripts/code that stopped working after this change, because I don't see any actual impact. I just don't get what's the problem with explicit path to cache, using already existing `tmp` dir. I just saw that `tmp/rector` is already there and used the same directory.

It's impossible to move it to "Docker setup scripts" since there are not any wrapper scripts where additional options could be added, also ECS and PHPStan does not have option for setting cache path on CLI level. The only way that could work would be adding some ugly workarounds for copying/extending configs in real time.

Comment (Wirone) [reply ecs.php]: @TomasVotruba ping, can you provide answer for my concerns above? Because I don't know if we can find some way to keep this explicit cache path and without that it does not make sense to run these tools inside container (full analysis every time).

Comment (TomasVotruba) [reply ecs.php]: Cache directories can be improved. The issue I have is with leaking this knowledge to all the configs we use :).
This should be handled on cache level itself.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -20,14 +20,33 @@ composer check-platform-reqs
...
-
 Then you can start working with the code :+1:
 
 <br>
 
 Do you want to **contribute a failing test**? [This tutorial will sow you how](https://github.com/rectorphp/rector/blob/main/docs/how_to_add_test_for_rector_rule.md)
 
+### Docker runtime
+
+*Note: using Docker for contributing is strongly discouraged, as it requires [extra knowledge of composer internals](https://github.com/composer/composer/issues/9368#issuecomment-718112361).*
```
Comment (Wirone) [inline CONTRIBUTING.md:31]: @TomasVotruba I've just rebased my changes on top of `main` and resolved conflicts. I've moved this information to this section.

However I strongly disagree with this 😃 Issue that you mentioned here is applicable only to CI process, as @seldaek mentioned. It does NOT affect local development (and `COMPOSER_ROOT_VERSION` is set inside container anyway). The only required thing for using Docker runtime is Docker knowledge and you really should NOT discourage people from using it. Just let people decide if they want pollute local machine with libraries or use Docker.

Comment (TomasVotruba) [reply CONTRIBUTING.md:31]: I've moved this after one contributor could not get through the `composer install` due to this :) 
I want to keep contributing simple and easy.

Comment (Wirone) [reply CONTRIBUTING.md:31]: So let's change it to discourage Docker usage only for people don't familiar with Docker. It does not have anything to Composer and Git since there's exact same repo inside container as on localhost (because it's mapped through volume). If developer knows how to use Docker properly there won't be problems with this approach.

PS. Can you point me to the `composer install` problem?

Comment (TomasVotruba) [reply CONTRIBUTING.md:31]: Here: https://github.com/rectorphp/rector/issues/7086#issuecomment-1092531861

**Comment by TomasVotruba** [comment]
Closing as the cache directories remain unresolved last  2 months and the complexity is to high to review properly.

If still needed, please reopen smaller PR with isolated changes in docs and docker config, so we can review and merge in matter of days as other PRs :) thanks

**Comment by Wirone** [comment]
From my point of view it was ready and I did not receive feedback what's actually wrong and how to change it, so I did not work on it. In terms of Docker runtime it is the only way to handle it properly. I will think about it, but it looks like you just don't share the idea of runtime encapsulation 😉

File: ecs.php (hunk 3 above)
Comment (TomasVotruba) [reply ecs.php]: This was not resolved. I don't know the solution, but it can't be here :)

Comment (Wirone) [reply ecs.php]: @TomasVotruba there's nothing to resolve IMHO, this change does not break anything and works for native usage too (all checks green!). It just ensures the common, stable bridge between local filesystem and mounted volume. Additionally, it's transparent to the users because they don't use cache directly.

File: CONTRIBUTING.md (hunk 5 above)
Comment (Wirone) [reply CONTRIBUTING.md:31]: [That's not a Docker runtime problem, that's developer's problem](https://github.com/rectorphp/rector/issues/7086#issuecomment-1166570839) (invalid usage, unfamiliarity with Docker etc.). I don't think he even executed that command in container.
